### PR TITLE
LangRef: Clarify llvm.minnum and llvm.maxnum about sNaN and signed zero

### PR DESCRIPTION
**Pull request by wzssyqa**
The documents claims that it ignores sNaN, while in the current code it may be different.

 - as the finally callback, it use libc call fmin(3)/fmax(3). while C23 clarifies that fmin(3)/fmax(3) should return NaN for sNaN vs NUM.
 - on some architectures, such as aarch64, it converts to `fmaxnm`, which returns qNaN for sNaN vs NUM.
 - on RISC-V (SPEC 2019+), it converts to `fmax`, which returns NUM for sNaN vs NUM.

Since we have introduced llvm.minimumnum and llvm.maximumnum, which follow IEEE 754-2019's minimumNumber/maximumNumber.

So, it's time for us to clarify llvm.minnum and llvm.maxnum. Since the final fallback of llvm.minnum and llvm.maxnum is
fmin(3)/fmax(3), so that it is reasonable to follow the behaviors of fmin(3)/fmax(3).

Although C23 clarified the behavior about sNaN and +0.0/-0.0:
     (NUM or NaN) vs sNaN -> qNaN
     +0.0 vs -0.0 -> either one of +0.0/-0.0
It is the same the IEEE754-2008's maxNUM and minNUM.
Not all implementation work as expected.
     
Since some architectures such as aarch64/MIPSr6/LoongArch, have instructions that implements +0.0>-0.0.
So Let's define llvm.minnum and llvm.maxnum to IEEE754-2008 with +0.0>-0.0.

The architectures without such instructions can implements `NSZ` flavor to speed up,
and the frontend, such as clang, can call them with `nsz` attribute.

### CODE REVIEW

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -16464,21 +16464,13 @@ type.
 
 Semantics:
 """"""""""
+Follows the IEEE754 2008 semantics for minNum, except for handling of
++0.0 vs -0.0. This matches the behavior of libm's fmin.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16468]: IEEE 754 has the same indifference to the result of `minNum(-0.0, +0.0)`, it's not a deviation from the IEEE 754 specification.

**Review by jcranmer-intel — COMMENTED** [review]
I'm not convinced this change is a good idea, as opposed to fixing the backends that implement it differently. LLVM has a general principle that sNaNs are treated more or less equivalent to qNaN, and this is the main case where sNaN inputs do very different things from qNaN inputs. The original text is pretty explicit that targets are supposed to use `llvm.canonicalize` to handle sNaN quieting if they're mapping to hardware that does distinguish between sNaN/qNaN, and I don't see any justification in this PR yet for why doing that is problematic.

**Comment by wzssyqa** [comment]
> I'm not convinced this change is a good idea, as opposed to fixing the backends that implement it differently. LLVM has a general principle that sNaNs are treated more or less equivalent to qNaN, and this is the main case where sNaN inputs do very different things from qNaN inputs. The original text is pretty explicit that targets are supposed to use `llvm.canonicalize` to handle sNaN quieting if they're mapping to hardware that does distinguish between sNaN/qNaN, and I don't see any justification in this PR yet for why doing that is problematic.

Yes. We should fix the backends, while before we fix them, we need a standard that what extractly to do.
Since in lots of cases that llvm.minnum will fallback to libc call, so that, aligning with the behavior of fmin(3) is the most reasonable
way.

For `sNaNs are treated more or less equivalent to qNaN`, it was reasonable before we have IEEE754-2019 and C23.
While C23 clarifies fmin(3)/fmax(3), and IEEE754-2019/C23 introduced minimumNumber.
It is time to do some clarification for fmin(3).
I don't think that it is a good idea to have some suprise to users with code like:
```
float f(float a, float b) {
   return fminf(a, b);
}
```
As users may read manual from libc, where states that fmax(N, sNaN) => NaN, while we (llvm) produces N.

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -16464,8 +16464,8 @@ type.
 
 Semantics:
 """"""""""
-Follows the IEEE754 2008 semantics for minNum, except for handling of
-+0.0 vs -0.0. This matches the behavior of libm's fmin.
+Follows the IEEE754 2008 semantics for minNum.
+This also matches the behavior of libm's fmin.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16468]: I actually think we should change the intrinsic to match the stricter 2019 signed zero behavior. The old behavior is equivalent to a stricter behavior with the nsz flag. 

This note about "libm's fmin" is too simple given the history of broken libms.

Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16468]: Since C23 doesn't require it and libc implementions doesn't do it.
For some architectures, such as X86/ARM-preV8/MIPS-prer6, it will need more work to archive +0.0>-0.0.
So I think that we should have a intrinsic just suite for libc's fmin/fmax.

Of course, for architectures that have FMINNUM_IEEE/FMAXNUM_IEEE, we can map FMINNUM/FMAXNUM to the _IEEE  ones, just as we did for ARMv8 and MIPSr6. 

I will add more clairification about the history of libm.

Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16468]: I you feel that the "IEEE754-2008 + +0.0>-0.0" is useful, we can export ISD::FMINNUM_IEEE as an new intrinsic.
Anyway, it will be another patch/PR.

**Comment by arsenm** [comment]
> I'm not convinced this change is a good idea, as opposed to fixing the backends that implement it differently.

I think we definitely need to do this. Taking the name of the IEEE operations and not implementing the same behavior is not an acceptable place to be. We will simplify the code, and save instructions by doing this. We now have minimumnum/maximumnum to provide the ignore-signalingness behavior.


> LLVM has a general principle that sNaNs are treated more or less equivalent to qNaN, and this is the main case where sNaN inputs do very different things from qNaN inputs. 

Yes, the IEEE behavior absurdly inverts the behavior for signaling nans so we do not have the luxury of simply ignoring them like in every other operation.

> The original text is pretty explicit that targets are supposed to use `llvm.canonicalize` to handle sNaN quieting if they're mapping to hardware that does distinguish between sNaN/qNaN, and I don't see any justification in this PR yet for why doing that is problematic.

This was me working around this absolute nonsense situation. The *only* reason it is this way is because of the OpenCL conformance test checked for signaling nans treated as quiet nans, and gnu libm's fmin/fmax had the ignore signaling behavior at the time. libm has been fixed, and OpenCL has a footnote on fmin/fmax that signaling nans "may" behave as quiet nans. 

The codegen for these was never consistent across targets. We never considered the host library versions for the buggy libm, and until recently I believe AMDGPU was the only target bothering with the canonicalize + ieee variant lowering. We have been wasting cycles on this for years, for no real reason. Signaling nans are not used in the real world. We should be able to just directly select this to the instruction that implements the IEEE 2008 behavior.

File: llvm/docs/LangRef.rst (hunk 2 above)
Comment (arsenm) [reply llvm/docs/LangRef.rst:16468]: > For some architectures, such as X86/ARM-preV8/MIPS-prer6, it will need more work to archive +0.0>-0.0.

Correct, these will need some kind of lowering code.

> So I think that we should have a intrinsic just suite for libc's fmin/fmax.

No, you can implement these with an nsz flag on the one intrinsic with the stricter definition. You should end up with the same codegen if you know the sign of zero is insignificant. The last thing we need is more min/max intrinsic variants.

> we can export ISD::FMINNUM_IEEE as an new intrinsic.

There should be no new intrinsic. FMINNUM_IEEE should just be the minnum behavior for the existing intrinsic

**Review by efriedma-quic — COMMENTED** [review]
Please add some mention of the interaction with FP arithmetic. (In particular, that arithmetic on an sNaN doesn't consistently produce a qNaN, so arithmetic feeding into a minnum/maxnum can produce inconsistent results.)

**Comment by wzssyqa** [comment]
> Please add some mention of the interaction with FP arithmetic. (In particular, that arithmetic on an sNaN doesn't consistently produce a qNaN, so arithmetic feeding into a minnum/maxnum can produce inconsistent results.)

qNaN is always produced in the following cases: 
1. In IEEE754-2008, qNaN is required for maxNUM and minNUM operations.
2. On architectures that implement these operation with hardware, such as MIPSr6, ARMv8, qNaN is also produced.
3. With the current glibc(2.25+) implementation, qNaN is also.

Yes, for older libc implementations (2.24-) qNaN vs sNaN may produce sNaN.

**Comment by efriedma-quic** [comment]
Say you write `fmax(x+0, y)`, and x is an snan.  The result is unpredictable unless you use constrained intrinsic.  If we optimize out the "+0", it produces qnan; if we don't, it returns 1.0.

**Comment by wzssyqa** [comment]
> Say you write `fmax(x+0, 1.0)`, and x is an snan. The result is unpredictable unless you use constrained intrinsic. If we optimize out the "+0", it produces qnan; if we don't, it returns 1.0.

done

**Review by jcranmer-intel — CHANGES_REQUESTED** [review]
This PR as it stands has numerous issues that need to be addressed.

At this point, we are now changing the behavior of the intrinsic with respect to signed zero handling, and changing the behavior with respect to sNaN. This means that the commit message is no longer accurate and needs to be adjusted.

The deeper problem is the text. The edge case behavior of these intrinsics is confusing, and it's not helped by the fact that the source materials tend to bury the behavior in the important edge cases so that you need to collate normative text from several different places to even understand what IEEE 754:2008's rules for minNum or C23's rules for fmin are. Because of that, the semantics really need to be upfront and clear about what the edge-case behavior is, and the history of changes so far have served to bury those details rather than illuminate them.

An example opening paragraph that would work better is this:
> Follows the IEEE-754 semantics for minNum, except that -0.0 < +0.0 for the purposes of this intrinsic. As for signaling NaNs, per the IEEE-754 semantics, if either operand is an sNaN, the result is always a qNaN. This matches the *recommended* behavior for the libm function `fmin`, although not all implementations have implemented these recommended behaviors.

After that opening paragraph, I think it's useful to keep the original second paragraph, except modified slightly to account for the changes to sNaN and -0.0/+0.0 semantics. And only after that does it make sense to start to discuss the interactions with `nsz`, LLVM's general sNaN-quieting rules, and the caution about implementing it properly.

(It also might be worth adding notes to the minnum/maxnum intrinsics that mininum/minimumnum intrinsics should probably be preferred for their non-distinction of sNaN/qNaN handling).

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -16464,21 +16464,32 @@ type.
...
+of this intrinsic. As for signaling NaNs, per the IEEE-754 semantics, if either operand
+is an sNaN, the result is always a qNaN. This matches the recommended behavior for the libm
+function fmin, although not all implementations have implemented these recommended behaviors.
+
+If either operand is a qNaN, returns the other non-NaN operand. Returns
+NaN only if both operands are NaN or either operand is sNaN.
 
-Follows the IEEE-754 semantics for minNum, except for handling of
-signaling NaNs. This match's the behavior of libm's fmin.
+If the operands compare equal, returns either one of the operands.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16475]: I'd delete this line, because the only time this happens are for -0/+0 (which is handled specially), decimal cohorts (which aren't supported by LLVM), or noncanonical values (which need their own discussion).

File: llvm/docs/LangRef.rst [hunk 4]
```diff
@@ -16464,21 +16464,32 @@ type.
...
+some applications like Clang, can call '``llvm.minnum.*``' with '``nsz``' attribute
+to archive the required behaivors of libm's fmin.
 
-Unlike the IEEE-754 2008 behavior, this does not distinguish between
-signaling and quiet NaN inputs. If a target's implementation follows
-the standard and returns a quiet NaN if either input is a signaling
-NaN, the intrinsic lowering is responsible for quieting the inputs to
-correctly return the non-NaN input (e.g. by using the equivalent of
-``llvm.canonicalize``).
+Some architecturs, such as ARMv8, LoongArch, MIPSr6, PowerPC/VSX, have the
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16481]: s/architecturs/architectures/

File: llvm/docs/LangRef.rst [hunk 5]
```diff
@@ -19407,12 +19430,12 @@ The '``llvm.vector.reduce.fmax.*``' intrinsics do a floating-point
...
-intrinsic. That is, the result will always be a number unless all elements of
-the vector are NaN. For a vector with maximum element magnitude 0.0 and
-containing both +0.0 and -0.0 elements, the sign of the result is unspecified.
+intrinsic.  If the intrinsic call has the ``nnan`` fast-math flag, then the
+operation can assume that NaNs are not present in the input vector.
 
-If the intrinsic call has the ``nnan`` fast-math flag, then the operation can
-assume that NaNs are not present in the input vector.
+It is deprecated, since the different order of inputs may produce different
+outputs, and it is hard to optimize with Vector or SIMD extensions.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:19437]: Nit: lower-case vector here (repeated for the other intrinsics).

File: llvm/include/llvm/CodeGen/ISDOpcodes.h [hunk 6]
```diff
@@ -1021,13 +1021,20 @@ enum NodeType {
   LRINT,
   LLRINT,
 
-  /// FMINNUM/FMAXNUM - Perform floating-point minimum or maximum on two
-  /// values.
+  /// FMINNUM/FMAXNUM - Perform floating-point minimum maximum on two values,
+  /// following IEEE-754 definitions.
```
Comment (jcranmer-intel) [inline llvm/include/llvm/CodeGen/ISDOpcodes.h:1025]: Add "except for signed zero behavior"

File: llvm/docs/LangRef.rst [hunk 7]
```diff
@@ -16464,21 +16464,32 @@ type.
...
+to archive the required behaivors of libm's fmin.
 
-Unlike the IEEE-754 2008 behavior, this does not distinguish between
-signaling and quiet NaN inputs. If a target's implementation follows
-the standard and returns a quiet NaN if either input is a signaling
-NaN, the intrinsic lowering is responsible for quieting the inputs to
-correctly return the non-NaN input (e.g. by using the equivalent of
-``llvm.canonicalize``).
+Some architecturs, such as ARMv8, LoongArch, MIPSr6, PowerPC/VSX, have the
+strictly same instructions; thus it is quite simple for these architectures.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16482]: s/the strictly same instructions/instructions that match these semantics exactly/.

File: llvm/docs/LangRef.rst [hunk 8]
```diff
@@ -16464,21 +16464,32 @@ type.
...
+For other architectures, the custom or expand methods may provide '``nsz``' flavor.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16483]: This isn't really a coherent description of what needs to be done for architectures without exact implementations of the semantics here.

File: llvm/docs/LangRef.rst [hunk 9]
```diff
@@ -16464,21 +16464,32 @@ type.
...
-signaling NaNs. This match's the behavior of libm's fmin.
+If the operands compare equal, returns either one of the operands.
 
-If either operand is a NaN, returns the other non-NaN operand. Returns
-NaN only if both operands are NaN. If the operands compare equal,
-returns either one of the operands. For example, this means that
-fmin(+0.0, -0.0) returns either operand.
+Returns -0.0 for +0.0 vs -0.0. libm doesn't require it, so that
+some applications like Clang, can call '``llvm.minnum.*``' with '``nsz``' attribute
+to archive the required behaivors of libm's fmin.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16479]: I'd reword the second sentence along these lines (unless anyone else has better wording suggestings):

> This behavior is more strict than the definition in C and IEEE 754, where either zero may be returned. To achieve the same permissiveness, one may use the `nsz` attribute on the intrinsic call.

File: llvm/docs/LangRef.rst [hunk 10]
```diff
@@ -16464,21 +16464,32 @@ type.
...
+For other architectures, the custom or expand methods may provide '``nsz``' flavor.
+
+Historically, libc returns NUM for NUM vs (sNaN or qNaN), and may return
+sNaN for qNaN vs sNaN. With the recent libc versions, libc follows IEEE754-2008:
+NUM vs sNaN -> qNaN; NUM vs qNaN -> NUM; qNaN vs sNaN -> qNaN; sNaN vs sNaN -> qNaN.
+
+Note that that arithmetic on an sNaN doesn't consistently produce a qNaN,
+so arithmetic feeding into a minnum can produce inconsistent results.
+Such as `fmin(sNaN+0.0, 1.0)` can produce qNaN or 1.0 depending on whether `+0.0`
+is optimized out.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16492]: How does this work instead?

> Note that, in LLVM IR, floating-point operations involving sNaNs may return an sNaN instead of a qNaN, so the result of a minnum following computation may change as a result of optimization. For example, `llvm.minnum(sNaN + 0.0, 1.0)` may return either `1.0` or a `qNaN`, depending on whether the `fadd` was optimized away.

File: llvm/docs/LangRef.rst [hunk 11]
```diff
@@ -16464,21 +16464,32 @@ type.
...
-NaN, the intrinsic lowering is responsible for quieting the inputs to
-correctly return the non-NaN input (e.g. by using the equivalent of
-``llvm.canonicalize``).
+Some architecturs, such as ARMv8, LoongArch, MIPSr6, PowerPC/VSX, have the
+strictly same instructions; thus it is quite simple for these architectures.
+For other architectures, the custom or expand methods may provide '``nsz``' flavor.
+
+Historically, libc returns NUM for NUM vs (sNaN or qNaN), and may return
+sNaN for qNaN vs sNaN. With the recent libc versions, libc follows IEEE754-2008:
+NUM vs sNaN -> qNaN; NUM vs qNaN -> NUM; qNaN vs sNaN -> qNaN; sNaN vs sNaN -> qNaN.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16487]: I really don't like this paragraph. It gives the strong impression that there is only one libc implementation that LLVM concerns itself with, which is of course far from the truth. I don't know that it adds anything here compared to the admonishment in the first paragraph that not all library implementations are always correct here.

File: llvm/docs/LangRef.rst (hunk 8 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16483]: I delete this line. and see bellow.

File: llvm/docs/LangRef.rst (hunk 10 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst]:
```
This behavior is more strict than the definition in C and IEEE 754, where either zero may be returned.
To achieve the same permissiveness, the backend may implement the nsz attribute, and one may use the nsz
attribute on the intrinsic call.
```

File: llvm/docs/LangRef.rst (hunk 11 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst]:
```
In the real libc worlds, the bebhaviors of fmin may quite different on sNaN and signed zero behaviors,
even in the same release of a single libm implemention. Such as in glibc 2.24, the Arm64 asm implemention
has different behaviour with the generic C implemention.
```

**Comment by wzssyqa** [comment]
@jcranmer-intel @arsenm ping

**Comment by wzssyqa** [comment]
@arsenm @jcranmer-intel ping

**Comment by wzssyqa** [comment]
@arsenm @jcranmer-intel any idea?

File: llvm/docs/LangRef.rst [hunk 12]
```diff
@@ -16464,21 +16464,29 @@ type.
...
+This behavior is more strict than the definition in C and IEEE 754, where either zero may be returned.
+To achieve the same permissiveness, the backend may implement the nsz attribute, and one may use the nsz
+attribute on the intrinsic call.
 
-If either operand is a NaN, returns the other non-NaN operand. Returns
-NaN only if both operands are NaN. If the operands compare equal,
-returns either one of the operands. For example, this means that
-fmin(+0.0, -0.0) returns either operand.
+Some architectures, such as ARMv8, LoongArch, MIPSr6, PowerPC/VSX, have instructions that match
+these semantics exactly; thus it is quite simple for these architectures.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16480]: It would be more useful, I think, to highlight the exact equivalent instructions on one or two instructions, and also highlight that, say, x86's `MINPS` instruction doesn't implement these semantics.

File: llvm/docs/LangRef.rst [hunk 13]
```diff
@@ -16464,21 +16464,29 @@ type.
...
 
-Unlike the IEEE-754 2008 behavior, this does not distinguish between
-signaling and quiet NaN inputs. If a target's implementation follows
-the standard and returns a quiet NaN if either input is a signaling
-NaN, the intrinsic lowering is responsible for quieting the inputs to
-correctly return the non-NaN input (e.g. by using the equivalent of
-``llvm.canonicalize``).
+In the real libc worlds, the bebhaviors of fmin may be quite different on sNaN and signed zero behaviors,
+even in the same release of a single libm implemention. Such as in glibc 2.24, the Arm64 asm implemention
+has different behaviour with the generic C implemention.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16484]: I don't know that we need this much detail on libm implementations of fmin.

File: llvm/docs/LangRef.rst [hunk 14]
```diff
@@ -16464,21 +16464,29 @@ type.
...
+
+Note that that arithmetic on an sNaN doesn't consistently produce a qNaN,
+so arithmetic feeding into a minnum can produce inconsistent results.
+Such as `fmin(sNaN+0.0, 1.0)` can produce qNaN or 1.0 depending on whether `+0.0`
+is optimized out.
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16489]: If you're introducing an example in a separate sentence, the proper English introductory phrase is not "Such as" but rather something like "For example,"

I think this example works better if you address it explicitly as two instructions, an fadd and the minnum call, and point out the difference if the fadd call is folded.

File: llvm/include/llvm/CodeGen/ISDOpcodes.h [hunk 15]
```diff
@@ -1021,13 +1021,20 @@ enum NodeType {
...
-  /// In the case where a single input is a NaN (either signaling or quiet),
-  /// the non-NaN input is returned.
+  /// These treat -0 as ordered less than +0, matching the behavior of IEEE-754
+  /// 2019's minimumNumber/maximumNumber.
   ///
-  /// The return value of (FMINNUM 0.0, -0.0) could be either 0.0 or -0.0.
+  /// Note that that arithmetic on an sNaN doesn't consistently produce a qNaN,
+  /// so arithmetic feeding into a minnum/maxnum can produce inconsistent
+  /// results. FMAXIMUN/FMINIMUM or FMAXIMUMNUM/FMINIMUMNUM may be better choice
+  /// for non-distinction of sNaN/qNaN handling.
```
[the remark this answers is not in the document]
Comment (wzssyqa) [reply llvm/include/llvm/CodeGen/ISDOpcodes.h:1037]: Yes, we are deprecating `FMINNUM_IEEE`, since we are claiming that `FMINNUM` is the same with `FMINNUM_IEEE`.

In future patch, we will remove `FMINNUM_IEEE`, but it will need a couple of patches.

File: llvm/docs/LangRef.rst [hunk 16]
```diff
@@ -16574,21 +16574,35 @@ type.
 
 Semantics:
 """"""""""
+Follows the IEEE-754 semantics for minNum, except that -0.0 < +0.0 for the purposes
+of this intrinsic. As for signaling NaNs, per the IEEE-754 semantics, if either operand
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16578]: This is the IEEE-754 behavior (just the 2019 version)

File: llvm/docs/LangRef.rst [hunk 17]
```diff
@@ -16574,21 +16574,35 @@ type.
...
+is an sNaN, the result is always a qNaN. This matches the recommended behavior for the libm
+function `fmin`, although not all implementations have implemented these recommended behaviors.
+
+If either operand is a qNaN, returns the other non-NaN operand. Returns
+NaN only if both operands are NaN or if either operand is sNaN.
+
+This behavior is more strict than the definition in C and IEEE 754, where either zero may be returned.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16585]: Stricter than IEEE-754 2008

File: llvm/docs/LangRef.rst [hunk 18]
```diff
@@ -16574,21 +16574,35 @@ type.
...
+To achieve the same permissiveness, the backend may implement the nsz attribute, and one may use the nsz
+attribute on the intrinsic call.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16587]: Use the nsz attribute

File: llvm/docs/LangRef.rst [hunk 19]
```diff
@@ -16574,21 +16574,35 @@ type.
...
+
+If the intrinsic is marked with the nsz attribute, then the effect is as in the definition in C
+and IEEE 754: the result of minnum(-0.0, +0.0) may be either -0.0 or +0.0.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16590]: Specify version and hyphenate

File: llvm/docs/LangRef.rst [hunk 20]
```diff
@@ -16625,20 +16639,34 @@ type.
 
 Semantics:
 """"""""""
-Follows the IEEE-754 semantics for maxNum except for the handling of
-signaling NaNs. This matches the behavior of libm's fmax.
+Follows the IEEE-754 semantics for minNum, except that -0.0 < +0.0 for the purposes
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16642]: Same as above, this changed between IEEE-754 2008 and 2019

**Review by arsenm — COMMENTED** [review]
Title and message should also reflect this changes the signed 0 behavior

File: llvm/docs/LangRef.rst (hunk 16 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16578]: The names of the operations are different in 2008 and 2019.
```   
2008 minNum
2019 minimumNumber
```

I don't think that it is a good idea to mention 2019 here, which will confuse the reader.
I will change IEEE-754 to IEEE754-2008 then to make it more clear.

File: llvm/docs/LangRef.rst (hunk 20 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16642]: I try to not use the words like `change` here.  In the IEEE754 wordings, 2019 removes `minNum/maxNum`, and introduces new `minimumNumber/maximumNumber`.

Comment (arsenm) [reply llvm/docs/LangRef.rst:16642]: The intrinsic now has the semantics as IEEE-754 2019 minimumNumber. The 2008 mention of minNum can be the footnote.

File: llvm/docs/LangRef.rst [hunk 21]
```diff
@@ -16574,21 +16574,35 @@ type.
 
 Semantics:
 """"""""""
+Follows the semantics of minNum in IEEE-754-2008, except that -0.0 < +0.0 for the purposes
+of this intrinsic. As for signaling NaNs, per the minNum semantics, if either operand
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16578]: Describe with 2019 name. The 2008 version can be mentioned as a historical footnote

File: llvm/docs/LangRef.rst [hunk 22]
```diff
@@ -23169,13 +23197,14 @@ result type. If only ``nnan`` is set then the neutral value is ``-Infinity``.
...
-unless all elements of the vector and the starting value are ``NaN``. For a
-vector with maximum element magnitude ``0.0`` and containing both ``+0.0`` and
-``-0.0`` elements, the sign of the result is unspecified.
+'``llvm.maxnum.*``' intrinsic).
 
 To ignore the start value, the neutral value can be used.
 
+It is deprecated, since the different order of inputs may produce different
+outputs, and it is hard to optimize with vector or SIMD extensions.
+Use '``llvm.vp.vector.reduce.fmaximum``' or '``llvm.vp.vector.reduce.fmaximumnum``' instead.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:23206]: This seems like a separate change? The vp intrinsic isn't really a drop in substitute, is it? If it is we can just delete the intrinsic and autoupgrade

File: llvm/docs/LangRef.rst [hunk 23]
```diff
@@ -16574,21 +16574,34 @@ type.
...
 
-Follows the IEEE-754 semantics for minNum, except for handling of
-signaling NaNs. This match's the behavior of libm's fmin.
 
-If either operand is a NaN, returns the other non-NaN operand. Returns
-NaN only if both operands are NaN. If the operands compare equal,
-returns either one of the operands. For example, this means that
-fmin(+0.0, -0.0) returns either operand.
+In the real libc worlds, the bebhaviors of fmin may be quite different on sNaN and signed zero behaviors,
+even in the same release of a single libm implemention.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16599]: I don't think this comment belongs in the langref. In theory the backend should be fixing this up to provide consistent behavior

File: llvm/docs/LangRef.rst [hunk 24]
```diff
@@ -16574,21 +16574,34 @@ type.
...
-Unlike the IEEE-754 2008 behavior, this does not distinguish between
-signaling and quiet NaN inputs. If a target's implementation follows
-the standard and returns a quiet NaN if either input is a signaling
-NaN, the intrinsic lowering is responsible for quieting the inputs to
-correctly return the non-NaN input (e.g. by using the equivalent of
-``llvm.canonicalize``).
+Note that arithmetic on an sNaN doesn't consistently produce a qNaN,
+so arithmetic feeding into a minnum can produce inconsistent results.
+For example, `maxnum(fadd(sNaN, 0.0), 1.0)` can produce qNaN or 1.0 depending on whether `fadd`
+is folded.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16604]: This should just link to the section on signaling nan semantics in the IR

File: llvm/docs/LangRef.rst [hunk 25]
```diff
@@ -16625,20 +16638,33 @@ type.
 
 Semantics:
 """"""""""
-Follows the IEEE-754 semantics for maxNum except for the handling of
-signaling NaNs. This matches the behavior of libm's fmax.
+Follows the semantics of maxinumNumber in IEEE-754-2019, except for signaling NaNs. If either operand
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16641]: Typo maxinumNumber. Same comments as above for minnum

File: llvm/docs/LangRef.rst (hunk 23 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16599]: Maybe it is useful to tell the backend that depending on libm is not safe.

File: llvm/docs/LangRef.rst [hunk 26]
```diff
@@ -16574,21 +16574,34 @@ type.
 
 Semantics:
 """"""""""
+Follows the semantics of mininumNumber in IEEE-754-2019, except for signaling NaNs. If either operand
+is an sNaN, the result is always a qNaN. This matches the recommended behavior for the libm
+function `fmin`, although not all implementations have implemented these recommended behaviors.
```
[the remark this answers is not in the document]
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16579]: Yes. We describe it with both 2008/2019. Wish it may be clearer..

**Comment by wzssyqa** [comment]
@arsenm ping.

File: llvm/docs/LangRef.rst [hunk 27]
```diff
@@ -16723,21 +16723,29 @@ type.
 
 Semantics:
 """"""""""
+Follows the semantics of minimumNumber in IEEE-754-2019, except for signaling NaNs. If either operand
+is an sNaN, the result is always a qNaN. This matches the recommended behavior for the libm
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16727]: I can't decide if this is the correct way to describe it, relative to IEEE. It's IEEE-754-2008 minNum with stronger -0 behavior. This is probably clearer, given this is where the intrinsic got its name from

File: llvm/docs/LangRef.rst [hunk 28]
```diff
@@ -16774,20 +16782,29 @@ type.
...
-Follows the IEEE-754 semantics for maxNum except for the handling of
-signaling NaNs. This matches the behavior of libm's fmax.
+Follows the semantics of maximumNumber in IEEE-754-2019, except for signaling NaNs. If either operand
+is an sNaN, the result is always a qNaN. This matches the recommended behavior for the libm
+function ``fmax``, although not all implementations have implemented these recommended behaviors.
+
+If either operand is a qNaN, returns the other non-NaN operand. Returns NaN only if both operands are
+NaN or if either operand is sNaN. Note that arithmetic on an sNaN doesn't consistently produce a qNaN,
+so arithmetic feeding into a maxnum can produce inconsistent results. For example,
+`maxnum(fadd(sNaN, 0.0), 1.0)` can produce qNaN or 1.0 depending on whether `fadd` is folded.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16792]:
```suggestion
`maxnum(fadd(sNaN, -0.0), 1.0)` can produce qNaN or 1.0 depending on whether `fadd` is folded.
```

**Review by arsenm — COMMENTED** [review]
Basically LGTM

File: llvm/docs/LangRef.rst [hunk 29]
```diff
@@ -16723,21 +16723,29 @@ type.
...
+function ``fmin``, although not all implementations have implemented these recommended behaviors.
+
+If either operand is a qNaN, returns the other non-NaN operand. Returns NaN only if both operands are
+NaN or if either operand is sNaN. Note that arithmetic on an sNaN doesn't consistently produce a qNaN,
+so arithmetic feeding into a minnum can produce inconsistent results. For example,
+`minnum(fadd(sNaN, 0.0), 1.0)` can produce qNaN or 1.0 depending on whether `fadd` is folded.
 
-Follows the IEEE-754 semantics for minNum, except for handling of
-signaling NaNs. This match's the behavior of libm's fmin.
+IEEE-754-2008 defines minNum, and it is removed in IEEE-754-2019. The behavior of this intrinsic is
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16735]: Nit: "it was removed"

File: llvm/docs/LangRef.rst [hunk 30]
```diff
@@ -16723,21 +16723,29 @@ type.
...
-NaN, the intrinsic lowering is responsible for quieting the inputs to
-correctly return the non-NaN input (e.g. by using the equivalent of
-``llvm.canonicalize``).
+Some architectures, such as ARMv8 (FMINNM), LoongArch (fmin), MIPSr6 (min.fmt), PowerPC/VSX (xsmindp),
+have instructions that match these semantics exactly; thus it is quite simple for these architectures.
+Some architectures have similiar ones while they are not exact equivalent. Such as x86 implements `MINPS`,
+which implements the semantics of C code `a<b?a:b`: NUM vs qNaN always return qNaN. `MINPS` can be used
+if `nsz` and `nnan` are given.
+
+In the real libc worlds, the bebhaviors of fmin may be quite different on sNaN and signed zero behaviors,
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:16747]: Spelling mistake here: it should be "behaviors", not "bebhaviors".

I'd also change the introductory for phrase to "For existing libc implementations"

File: llvm/docs/LangRef.rst (hunk 27 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16727]: I also prefer describe it with IEEE-754-2008 as base.

**Comment by wzssyqa** [comment]
> Basically LGTM

OK for trunk now?

**Review by arsenm — APPROVED** [review]
LGTM, thanks for persisting on this

**Comment by topperc** [comment]
If these intrinsics now define the order of -0.0/+0.0, but the fmax and fmin library calls don't, how are we allowed to expand these intrinsics the fmax/fmin libcalls?

**Comment by wzssyqa** [comment]
> If these intrinsics now define the order of -0.0/+0.0, but the fmax and fmin library calls don't, how are we allowed to expand these intrinsics the fmax/fmin libcalls?

Yes. You are right. It is a bug yet not being fixed.

**Comment by nunoplopes** [comment]
Please see https://github.com/llvm/llvm-project/issues/138303
